rfd/0185-k8s-access-non-cert-routing.md

```diff
@@ -0,0 +1,247 @@
+    tls-server-name: kube-teleport-proxy-alpn.example.teleport.sh # No-change
+  name: example-example
+```
+
+#### Backwards/Forwards Compatibility
+
+By retaining the existing behaviour and endpoints within the Kube Forwarder,
+older clients using a kubeconfig and certificate generated with the
+`KubernetesCluster` attribute will continue to function.
+
+- VX.Y.0:
+  - Introduce new Kube Forwarder behaviour.
+  - Introduce opt-in `kubernetes/v2` `tbot` output. This will allow
+    users to opt-in to the new behaviour once they are satisfied that their
+    Auth Server, Proxy and Kubernetes Agents have been upgraded to the required
+    version.
+- V(X+1).0.0:
+  - Switch `tsh` to new behaviour. No longer requests the signing of a
+    certificate with the `KubernetesCluster` attribute and uses standard user
+    identity to authenticate.
+  - Switch `tbot` `kubernetes/v1` output to use `kubernetes/v2` implementation
+    under the hood.
+- V(X+2).0.0:
+  - It is now "safe" to remove the old attribute based routing from the Kube
+    Forwarder. We may wish to delay this further if we have concerns about
+    older client compatability.
+
+#### Client Compatability
+
+Kubernetes API clients have been known to vary greatly in their support for 
+more unusual configuration. We have witnessed this in the past when it came to
+supporting the ability to explicitly provide the SNI to use when making
+requests.
+
+Therefore, we should ensure that Kubernetes clients and SDKs do support
+providing a base path as part of the server address (e.g they do not strip off
+the path and only use the hostname/port).
+
+The key customer we are working with has specified that they use the following
+Kubernetes clients:
+
+- `kubectl`
+- Python Kubernetes SDK
+
+Both of these clients correctly handle a base-path as configured in Kubeconfig:
+
+```yaml
+clusters:
+  - cluster:
+      server: http://127.0.0.1:8888/example/base/path
+    name: test-base-path
+```
+
+#### Analytics
+
+It would no longer be possible to track usage of Kubernetes Access for analytics
+purposes via the certificate generation event. However, we have other events
+that indicate Kubernetes Access usage (e.g an event per request), that are
+better suited for this purpose - so this seems acceptable.
+
+#### TBot UX
+
+The `tbot` configuration will be extended with a new `kubernetes/v2` service
+type that will allow you to specify multiple kubernetes clusters to generate
+within a single Kubeconfig file.
+
+To select which clusters to include within the Kubeconfig, the name or labels
+of the clusters can be specified.
+
+```yaml
+- type: kubernetes/v2
+  destination:
+    type: directory
+    path: /opt/machine-id
+  # Each selector can either be a name or a set of labels. If multiple selectors
+  # are provided, then the output will contain the union of the results of the
+  # selectors.
+  selectors:
+    - name: my-cluster
```

Comment:
Will this generate a separate kubeconfig for each cluster, or a single configuration that includes all clusters?

I prefer the latter, but it might be worth mentioning or offering both options. We should also pay attention to context and file naming. By default, `tsh` creates context entries using the format `{{TeleportCluster}}-{{KubeCluster}}`, but this can be customized by providing a template with the context override flag. It would be useful to allow users to adjust these settings to customize their outputs.

Reply:
Yeah - following discussion with [CUSTOMER] - I figure it's easier to give an option between a single kubeconfig and a kubeconfig per cluster. I'm sure one you get into the thousands of clusters range, certain implementations of the Kube SDK will probably dislike the size of the config file.

> By default, tsh creates context entries using the format {{TeleportCluster}}-{{KubeCluster}}, but this can be customized by providing a template with the context override flag. It would be useful to allow users to adjust these settings to customize their outputs.

This seems reasonable - we use the same context structure in tbot as well atm fwiw.